homeassistant/components/modern_forms/fan.py

```diff
@@ -0,0 +1,195 @@
+    OPT_SPEED,
+    SERVICE_CLEAR_FAN_SLEEP_TIMER,
+    SERVICE_SET_FAN_SLEEP_TIMER,
+)
+
+
+async def async_setup_entry(
+    hass: HomeAssistantType, config_entry: ConfigEntry, async_add_entities: Callable
+) -> None:
+    """Set up a Modern Forms platform from config entry."""
+
+    coordinator: ModernFormsDataUpdateCoordinator = hass.data[DOMAIN][
+        config_entry.entry_id
+    ]
+
+    platform = entity_platform.async_get_current_platform()
+
+    platform.async_register_entity_service(
+        SERVICE_SET_FAN_SLEEP_TIMER,
+        {
+            vol.Required(ATTR_SLEEP_TIME): vol.All(
+                vol.Coerce(int), vol.Range(min=1, max=1440)
+            ),
+        },
+        "async_set_fan_sleep_timer",
+    )
+
+    platform.async_register_entity_service(
+        SERVICE_CLEAR_FAN_SLEEP_TIMER,
+        {},
+        "async_clear_fan_sleep_timer",
+    )
+
+    update_lights = partial(
+        async_update_fan, config_entry, coordinator, {}, async_add_entities
+    )
+    coordinator.async_add_listener(update_lights)
+    update_lights()
+
+
+class ModernFormsFanEntity(FanEntity, ModernFormsDeviceEntity):
+    """Defines a Modern Forms light."""
+
+    SPEED_RANGE = (1, 6)  # off is not included
+
+    def __init__(
+        self, entry_id: str, coordinator: ModernFormsDataUpdateCoordinator
+    ) -> None:
+        """Initialize Modern Forms light."""
+        super().__init__(
+            entry_id=entry_id,
+            coordinator=coordinator,
+            name=f"{coordinator.data.info.device_name} Fan",
+            icon="mdi:fan",
+        )
+
+    @property
+    def unique_id(self) -> str:
+        """Return the unique ID for this light."""
+        return f"{self.coordinator.data.info.mac_address}_fan"
+
+    @property
+    def supported_features(self) -> int:
+        """Flag supported features."""
+        return SUPPORT_DIRECTION | SUPPORT_SET_SPEED
+
+    @property
+    def percentage(self) -> int | None:
+        """Return the current speed percentage."""
+        percentage = 0
+        if bool(self.coordinator.data.state.fan_on):
+            percentage = ranged_value_to_percentage(
+                self.SPEED_RANGE, self.coordinator.data.state.fan_speed
+            )
+        return percentage
+
+    @property
+    def current_direction(self) -> str:
+        """Return the current direction of the fan."""
+        return self.coordinator.data.state.fan_direction
+
+    @property
+    def speed_count(self) -> int:
+        """Return the number of speeds the fan supports."""
+        return int_states_in_range(self.SPEED_RANGE)
+
+    @property
+    def is_on(self) -> bool:
+        """Return the state of the fan."""
+        return bool(self.coordinator.data.state.fan_on)
+
+    @modernforms_exception_handler
+    async def async_set_direction(self, direction: str) -> None:
+        """Set the direction of the fan."""
+        await cast(ModernFormsDataUpdateCoordinator, self.coordinator).modernforms.fan(
```

Comment:
Can this be typed in the base class instead?

Reply:
Yeah moved this higher